Add enable_feedforward parameter

pascalau opened the pull request:
Added a parameter to enable feedforward of the controller during configuration of the controller.
#1270 also mentions that there is currently no option to enable the feedforward gain with the exception of the service which can be unreliable to do from a launchfile.
Another option would be to set the state depending on the value of the feedforward gain but this could cause unexpected behaviour for users that have the value set but expect current behaviour.

christophfroehlich commented on pid_controller/src/pid_controller.yaml:
```diff
@@ -43,6 +43,11 @@ pid_controller:
     default_value: false,
     description: "Use external states from a topic instead from state interfaces."
   }
+  enable_feedforward: {
+    type: bool,
+    default_value: false,
+    description: "Enables feedforward gain on start."
+  }
```
I suggest renaming the parameter to `initial_state_feedforward_control` or similar, and set `read_only: true` for this parameter.

```suggestion
  initial_state_feedforward_control: {
    type: bool,
    default_value: false,
    read_only: false,
    description: "Enables feedforward control on startup."
  }
```

christophfroehlich reviewed (changes requested):
First of all, thanks for your contribution.

With the current implementation, the parameter is only used at configure() transition. But from the naming and the fact that it is not read_only, I can imagine that a user wants to use this parameter to change the behavior while the controller is active. (what would be easily possible, but would make the service `set_feedforward_control` obsolete).

christophfroehlich commented:
I'll bring that to tomorrow's WG meeting, let's see what the original author @destogl thinks about that

christophfroehlich commented:
We came to the conclusion that we can simplify this and remove the service at all. But we shouldn't do that on jazzy maybe, so we have to do that in two steps:
- Add deprecation warnings to this PR for the service call and the new enable parameter (e.g., _will be removed for the upcoming ROS 2 Kilted Kaiju release_)
- Create a new PR removing the stuff, which will be held until we branch for jazzy (somewhen before the K-turtle release).

pascalau commented:
How exactly should I add the deprecation warning?
What I would do:
- Error in rosout
- Deprecation warning in userdoc

Are there any other ways I am missing? Since the interaction of the user is via the service call there are no functions that can be marked deprecated.

pascalau commented:
I added deprecation warnings as outlined and also removed the enum `feedforward_mode`. Currently the control_mode_ variable is still needed since both the parameter and the service need to control the behaviour. After deprecating the service one only needs to change the if statement to directly use the parameter instead of the `control_mode_` variable and remove the unit test.

christophfroehlich reviewed (changes requested):
Thanks for the follow-up. My suggestions:

- If we remove the `feedforward_mode_type` enum, then we should rename the `control_mode_` variable to be more descriptive. `feedforward_mode_enabled_` for example.
- We should not use the `enable_feedforward` parameter later than on_configure. In the future, we want to remove this as well and just set the behavior by setting the `feedforward_gain` to zero or nonzero. If we use this parameter as suggested now, then we again have to deprecate it. Let it just control the behavior at startup (a temporary fix), and mark it already as deprecated if set to false (which is the default value). So users will switch to use the `feedforward_gain` directly.

Does that make sense for you?

pascalau commented on pid_controller/doc/userdoc.rst:
```diff
@@ -87,6 +90,9 @@ List of parameters
 =========================
 .. generate_parameter_library_details:: ../src/pid_controller.yaml
 
+.. warning::
+   The parameter ``enable_feedforward`` is being deprecated in favour of setting the ``feedforward_gain`` parameter to a non-zero value.
+   This might cause different behaviour if currently the ``feedforward_gain`` is set to a non-zero value and not activated.
```
Let's say someone has a configuration file with a non-zero feedforward_gain and never bothered to turn on the mode via the service. This would be different behaviour to what he previously expected.

christophfroehlich replied:
ok, but this not the case yet, that's why I was confused

pascalau replied:
I think I clarified it that it will currently work as expected but will change in the future.

saikishor commented on pid_controller/src/pid_controller.cpp:
```diff
@@ -96,6 +96,8 @@ void PidController::update_parameters()
   }
   params_ = param_listener_->get_params();
+
+  feedforward_mode_enabled_.writeFromNonRT(params_.enable_feedforward);
```
This one is missing when the feedforward is modified online, so we will have to add writeFromRT to update method. What do you think?

saikishor replied:
Sorry all good, I see that's what it is doing
